### PR TITLE
simplify some of the JTDSchema

### PR DESCRIPTION
**Pull request by erikbrinkman**
**What issue does this pull request resolve?**

The JTD schema was a little complex, it still is, but arguably less so.

**What changes did you make?**

This puts everything behind a guard for that schema type, and then instead of having things be unions of everything, it does sequential checks

**Is there anything that requires more attention while reviewing?**

It should be relatively straightforward.

### CODE REVIEW

File: lib/types/jtd-schema.ts [hunk 1]
```diff
@@ -28,11 +31,35 @@ type EnumString<T> = [T] extends [never]
...
+
+/** true if the type is a properties type */
+type IsProperties<T> = false extends IsUnion<Exclude<T, null>>
+  ? null extends EnumString<keyof Exclude<T, null>>
+    ? false
+    : true
+  : false
+
+/** true if type is discriminator type */
+type IsDiscriminator<T> = true extends IsUnion<Exclude<T, null>>
```
Comment (epoberezkin) [inline lib/types/jtd-schema.ts]: It is possibly to further refactor into something like

```
IsRecord<T, Union extends boolean> = Union extends IsUnion<Exclude<T, null>>
   ? null extends EnumString<keyof Exclude<T, null>>
     ? false
     : true
   : false
```

(spent some time looking for the difference between the two:)

File: lib/types/jtd-schema.ts [hunk 2]
```diff
@@ -48,94 +75,87 @@ export type JTDSchemaType<T, D extends Record<string, unknown> = Record<string,
...
+        : never
+      : // values
+      true extends IsValues<T>
+      ? T extends Record<string, infer V>
+        ? {
+            values: JTDSchemaType<V>
+          }
+        : never
+      : // properties
+      true extends IsProperties<T>
```
Comment (epoberezkin) [inline lib/types/jtd-schema.ts]: ... and then if you like refactoring idea this would be `IsRecord<T, false>`...

File: lib/types/jtd-schema.ts [hunk 3]
```diff
@@ -48,94 +75,87 @@ export type JTDSchemaType<T, D extends Record<string, unknown> = Record<string,
...
-      | (false extends IsUnion<Exclude<T, null>>
-          ? true extends TypeEquality<keyof Exclude<T, null>, string>
-            ? T extends Record<string, infer V>
-              ? {
-                  values: JTDSchemaType<V>
+              }) & {
+            additionalProperties?: boolean
+          }
+      : // discriminator
+      true extends IsDiscriminator<T>
```
Comment (epoberezkin) [inline lib/types/jtd-schema.ts]: ... and this `IsRecord<T, true>`

Or they could stay separate, defined via `IsRecord`.

**Comment by epoberezkin** [comment]
Great - let me know what you think - happy to merge as is or with refactoring.

**Review by epoberezkin — DISMISSED** [review]

File: lib/types/jtd-schema.ts (hunk 1 above)
Comment (erikbrinkman) [reply lib/types/jtd-schema.ts]: yeah, I like the brevity!

**Review by erikbrinkman — COMMENTED** [review]
yeah, I 'like the brevity. I'll make that change

**Review by epoberezkin — APPROVED** [review]

**Pull request merged**